### PR TITLE
Fix error handling in frontend and backend

### LINKED ISSUE #63: Error handling
## Description
The error handling is questionable at several places in the application. We should display more precise error messages and also make use of the error and loading states of the Swr hooks.

### Backend
- [ ] self-promote/demote: user not authorized
- [ ] request closed group: user not authorized
- [ ] reauth without valid access token: 500

### Frontend:
- [ ] inviteUser function in components/group/Member.tsx
- [ ] subscribe/unsubscribe function in componentes/input/NotificationButton.tsx
- [ ] add Group, Deck: naming restrictions (minimum number of characters, special characters)

### PR DESCRIPTION
**Pull request by TomRomeo**
## Description
This PR clarifies error messages in the backend and improves error display in the frontend with new toast library

## Resolved Issues
resolves #63

### CODE REVIEW

**Review by Julian702 — COMMENTED** [review]
Please fix the sonarcloud issues :)

File: backend/pkg/helper/security.go [hunk 1]
```diff
@@ -18,8 +18,17 @@ func IsAuthorized(groupRole pbCommon.GroupRole, requiredRole pbCommon.GroupRole)
...
 	if !pattern.MatchString(name) {
-		return NewMicroInvalidUserNameFormatErr(clientID)
+		return NewMicroInvalidNameFormatErr(clientID)
 	}
 	logger.Infof("Name %s is valid", name)
 	return nil
 }
+
+func CheckForValidPassword(password string, clientID ClientID) error {
+	logger.Infof("Check password for valid format")
```
Comment (darmiel) [inline backend/pkg/helper/security.go:28]: debug leftovers

**Review by darmiel — APPROVED** [review]
ta

File: frontend/components/input/NotificationButton.tsx [hunk 2]
```diff
@@ -40,30 +42,48 @@ export const NotificationButton = ({
...
+							className="h-full"
+							delay={0}
+							theme="simple"
+						/>
+						<span className="sr-only">Loading...</span>
+					</output>
+					<span>
+						<Trans>Loading...</Trans>
+					</span>
+				</div>
```
Comment (Julian702) [inline frontend/components/input/NotificationButton.tsx:80]:
```suggestion
				<div className="flex items-center space-x-2">
					<LoadingSpinner
						className="h-full"
						delay={0}
						theme="simple"
					/>
					<Trans>Loading...</Trans>
				</div>
```

File: frontend/components/navigation/Navbar.tsx [hunk 3]
```diff
@@ -44,7 +45,8 @@ export const Navbar = ({ className = "" }: NavbarProps) => {
 						const response = await authedFetch(logoutRoute, {
 							method: "POST",
 						});
-						if (response?.ok) {
+						if (response.ok) {
+							toast.success(t`Logged out`, { id: "logoutToast" });
```
Comment (Julian702) [inline frontend/components/navigation/Navbar.tsx:49]: We should agree on a consistent id format for all toasts.

File: frontend/pages/login.tsx [hunk 4]
```diff
@@ -251,12 +253,17 @@ export default function Page() {
 		) {
 			return;
 		}
+		const toastID = toast.loading(t`Logging in...`, { id: "loginToast" });
```
Comment (Julian702) [inline frontend/pages/login.tsx:256]: s.a.

File: backend/pkg/helper/errors.go [hunk 5]
```diff
@@ -26,111 +26,122 @@ var (
...
 
 func NewMicroAlreadyInvitedErr(id ClientID) error {
-	return microErrors.BadRequest(string(id), "user already invited")
+	return microErrors.BadRequest(string(id), "User already invited")
 }
 
-func NewMicroHashingFailedErr(id ClientID) error {
-	return microErrors.InternalServerError(string(id), "error while hashing password")
+func NewMicroCantModifyGroupAdminErr(id ClientID) error {
+	return microErrors.Unauthorized(string(id), "You can not modify group admins")
```
Comment (4KevR) [inline backend/pkg/helper/errors.go:93]: Maybe `cannot`?

File: backend/pkg/helper/errors.go [hunk 6]
```diff
@@ -26,111 +26,122 @@ var (
...
 }
 
-func NewMicroInvalidEmailOrPasswordErr(id ClientID) error {
-	return microErrors.BadRequest(string(id), "invalid email or password")
+func NewMicroCantKickGroupAdminErr(id ClientID) error {
+	return microErrors.Unauthorized(string(id), "You can not kick group admins")
```
Comment (4KevR) [inline backend/pkg/helper/errors.go:97]: s.a.

File: backend/pkg/helper/errors.go [hunk 7]
```diff
@@ -26,111 +26,122 @@ var (
...
 }
 
 func NewMicroDeckAlreadyFavoriteErr(id ClientID) error {
 	return microErrors.BadRequest(string(id), "This Deck is already your favorite")
 }
 
+func NewFiberReauthenticateError() error {
+	return fiber.NewError(fiber.StatusUnauthorized, "Please re-authenticate")
+}
 func NewFiberBadRequestErr(detail string) error {
```
Comment (4KevR) [inline backend/pkg/helper/errors.go:127]: missing newline

File: frontend/util/api.ts [hunk 8]
```diff
@@ -59,18 +61,35 @@ export async function submitForm(
...
 
 export async function deleteUser() {
+	const toastID = toast.loading(t`Deleting user...`);
 	const response = await deleteRequest(userRoute);
+	if (response.ok) {
+		toast.success(t`User deleted`, { id: toastID });
+	} else {
+		const error = await response.text();
+		toast.error(error, { id: toastID });
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:81]: Have seen this multiple times. Maybe refactor this to a util function?

**Review by Julian702 — CHANGES_REQUESTED** [review]
Great stuff, just some implementation details we should check

File: backend/services/collaboration/handler/collaboration.go [hunk 9]
```diff
@@ -433,9 +433,6 @@ func (e *Collaboration) ModifyGroupUserRequest(
 func (e *Collaboration) KickGroupUser(ctx context.Context, req *pbCommon.GroupModUserRequest, rsp *pbCommon.Success) error {
 	logger.Infof("Received Collaboration.KickGroupUserRequest request: %v", req)
 
-	if req.UserID == req.ModUserID {
-		return helper.NewMicroNotAuthorizedErr(helper.CollaborationServiceID)
-	}
```
Comment (4KevR) [inline backend/services/collaboration/handler/collaboration.go:438]: Why was this check removed? Should it be possible that a user kicks itself? I thought we have the `LeaveGroup` method for that.

File: backend/services/user/handler/user.go [hunk 10]
```diff
@@ -26,7 +26,10 @@ func New(s store.UserStore, cS pbCollaboration.CollaborationService) *User {
 }
 
 func (e *User) Register(ctx context.Context, req *pbCommon.User, rsp *pbCommon.Success) error {
-	logger.Infof("Received User.Register request: email: %v", req.UserEmail)
+	logger.Infof("Received User.Register request")
```
Comment (4KevR) [inline backend/services/user/handler/user.go:29]: Is this a conflicting change in regard to #240?

File: backend/services/user/handler/user.go [hunk 11]
```diff
@@ -117,11 +120,14 @@ func (e *User) DeleteUser(ctx context.Context, req *pbCommon.User, rsp *pbCommon
 }
 
 func (e *User) Login(ctx context.Context, req *pbCommon.User, rsp *pbCommon.User) error {
-	logger.Infof("Received User.Login request: email: %v", req.UserEmail)
+	logger.Infof("Received User.Login request")
```
Comment (4KevR) [inline backend/services/user/handler/user.go:123]: s.a.

File: frontend/pages/login.tsx [hunk 12]
```diff
@@ -51,7 +52,7 @@ export default function Page() {
 			<NextSeo
 				title={_(msg`Kioku | Login or register for Kioku!`)}
 				description={_(
-					msg`Register today and start using the free flashcard application together with your friends. Simply create new decks or import existing decks from Anki and collaborate in groups.`
+					msg`Register today and start using the free flashcard application together with your friends. Simply create new decks or import existing decks from Anki and collaborate in groups.`,
```
Comment (4KevR) [inline frontend/pages/login.tsx:55]: see general comment

**Review by 4KevR — COMMENTED** [review]
Looks great! Did not check in local build, I have some comments beforehand. Also, in some places, strange occurrences of commas appeared (see last comment), was this intended?

File: backend/services/user/handler/user.go (hunk 10 above)
Comment (TomRomeo) [reply backend/services/user/handler/user.go:29]: It is a similar change, however that should not lead to problems down the line. :)

File: backend/services/user/handler/user.go (hunk 11 above)
Comment (TomRomeo) [reply backend/services/user/handler/user.go:123]: as explained above

File: frontend/components/input/NotificationButton.tsx [hunk 13]
```diff
@@ -77,7 +95,15 @@ export const NotificationButton = ({
...
 			setSubscriptionId(await saveSubscription(subscription));
 			mutate(notificationsRoute);
-		} catch (err) {}
+			setTimeout(() => {
+				setLoading(false);
+			}, 500);
+		} catch (err) {
+			if (err instanceof Error) {
+				setLoading(false);
+				toast.error(err.message, { id: "notification-error" });
```
Comment (Julian702) [inline frontend/components/input/NotificationButton.tsx:104]: Use standardized format

File: frontend/components/navigation/Navbar.tsx [hunk 14]
```diff
@@ -44,7 +45,10 @@ export const Navbar = ({ className = "" }: NavbarProps) => {
 						const response = await authedFetch(logoutRoute, {
 							method: "POST",
 						});
-						if (response?.ok) {
+						if (response.ok) {
+							toast.success(t`Logged out`, {
+								id: "logoutToastID",
+							});
```
Comment (Julian702) [inline frontend/components/navigation/Navbar.tsx:51]: Is there a reason not to use the new `handleWithToast` function?

File: frontend/pages/login.tsx [hunk 15]
```diff
@@ -251,12 +253,17 @@ export default function Page() {
...
 			emailInput.current,
 			passwordInput.current,
 		]);
 		if (response.ok) {
+			toast.success(t`Logged in!`, { id: toastID });
 			router.push("/");
+		} else {
+			const error = await response.text();
+			toast.error(error, { id: toastID });
 		}
```
Comment (Julian702) [inline frontend/pages/login.tsx:267]: Here as well

File: frontend/pages/login.tsx [hunk 16]
```diff
@@ -271,14 +278,21 @@ export default function Page() {
...
 			passwordInput.current,
 		]);
 		if (response.ok) {
+			toast.success(t`Account created!`, { id: toastID });
 			setLogin(true);
 			emailInput.current?.focus();
+		} else {
+			const error = await response.text();
+			toast.error(error, { id: toastID });
 		}
```
Comment (Julian702) [inline frontend/pages/login.tsx:296]: And here again

File: frontend/util/api.ts [hunk 17]
```diff
@@ -94,13 +112,17 @@ async function groupInvitation(
...
+			invitationGroupRoute(groupID),
+			JSON.stringify({
+				invitedUserEmail: userEmail,
+			})
+		)
 	);
-	if (response?.ok) mutateAll(groupMemberRoutes(groupID));
+	if (response.ok) {
+		mutateAll(groupMemberRoutes(groupID));
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:125]:
```suggestion
	if (response.ok) mutateAll(groupMemberRoutes(groupID));
```

File: frontend/util/api.ts [hunk 18]
```diff
@@ -119,14 +141,15 @@ async function groupRequest(
...
-	if (response?.ok)
+	const response = await handleWithToast(request(requestGroupRoute(groupID)));
+	if (response.ok) {
 		mutateAll([
 			invitationsUserRoute,
 			groupsRoute,
 			groupRoute(groupID),
 			...groupMemberRoutes(groupID),
 		]);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:152]:
```suggestion
	if (response.ok)
		mutateAll([
			invitationsUserRoute,
			groupsRoute,
			groupRoute(groupID),
			...groupMemberRoutes(groupID),
		]);
```

File: frontend/util/api.ts [hunk 19]
```diff
@@ -58,19 +60,31 @@ export async function submitForm(
...
 
 export async function modifyUser(inputs: HTMLInputElement[]) {
-	const response = await submitForm(userRoute, inputs, putRequest);
-	if (response?.ok) mutate(userRoute);
+	const response = await handleWithToast(
+		submitForm(userRoute, inputs, putRequest)
+	);
+	if (response?.ok) {
+		mutate(userRoute);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:68]:
```suggestion
	if (response?.ok) mutate(userRoute);
```

File: frontend/util/api.ts [hunk 20]
```diff
@@ -58,19 +60,31 @@ export async function submitForm(
...
 
 export async function createGroup(inputs: HTMLInputElement[]) {
-	const response = await submitForm(groupsRoute, inputs, postRequest);
-	if (response?.ok) mutate(groupsRoute);
+	const response = await handleWithToast(
+		submitForm(groupsRoute, inputs, postRequest)
+	);
+	if (response.ok) {
+		mutate(groupsRoute);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:87]:
```suggestion
	if (response.ok) mutate(groupsRoute);
```

File: frontend/util/api.ts [hunk 21]
```diff
@@ -83,8 +97,12 @@ export async function modifyGroup(
...
 ) {
 	const route = groupRoute(groupID);
-	const response = await putRequest(route, JSON.stringify(body));
-	if (response?.ok) mutate(route);
+	const response = await handleWithToast(
+		putRequest(route, JSON.stringify(body))
+	);
+	if (response.ok) {
+		mutate(route);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:105]:
```suggestion
	if (response.ok) mutate(route);
```

File: frontend/util/api.ts [hunk 22]
```diff
@@ -141,27 +164,49 @@ export async function declineGroupInvitation(groupID: string) {
...
 
 export async function deleteMember(groupID: string, userID: string) {
-	const response = await deleteRequest(groupMemberRoute(groupID, userID));
-	if (response?.ok) mutate(groupMembersRoute(groupID));
+	const response = await handleWithToast(
+		deleteRequest(groupMemberRoute(groupID, userID))
+	);
+	if (response.ok) {
+		mutate(groupMembersRoute(groupID));
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:172]:
```suggestion
	if (response.ok) mutate(groupMembersRoute(groupID));
```

File: frontend/util/api.ts [hunk 23]
```diff
@@ -141,27 +164,49 @@ export async function declineGroupInvitation(groupID: string) {
...
-	if (response?.ok) mutate(groupsRoute);
+	const response = await handleWithToast(
+		deleteRequest(groupMembersRoute(groupID)),
+		"leaveGroupToastID",
+		t`Leaving group...`,
+		t`Left group`
+	);
+	if (response.ok) {
+		mutate(groupsRoute);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:185]:
```suggestion
	if (response.ok) mutate(groupsRoute);
```

File: frontend/util/api.ts [hunk 24]
```diff
@@ -141,27 +164,49 @@ export async function declineGroupInvitation(groupID: string) {
...
-	if (response?.ok) mutate(groupsRoute);
+	const response = await handleWithToast(
+		deleteRequest(groupRoute(groupID)),
+		"deleteGroupToastID",
+		t`Deleting group...`,
+		t`Group deleted`
+	);
+	if (response.ok) {
+		mutate(groupsRoute);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:198]:
```suggestion
	if (response.ok) mutate(groupsRoute);
```

File: frontend/util/api.ts [hunk 25]
```diff
@@ -141,27 +164,49 @@ export async function declineGroupInvitation(groupID: string) {
...
 export async function createDeck(inputs: HTMLInputElement[], groupID: string) {
 	const route = decksRoute(groupID);
-	const response = await submitForm(route, inputs, postRequest);
-	if (response?.ok) mutate(route);
+	const response = await handleWithToast(
+		submitForm(route, inputs, postRequest)
+	);
+	if (response.ok) {
+		mutate(route);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:209]:
```suggestion
	if (response.ok) mutate(route);
```

File: frontend/util/api.ts [hunk 26]
```diff
@@ -174,14 +219,25 @@ export async function modifyDeck(
...
 ) {
 	const route = deckRoute(deckID);
-	const response = await putRequest(route, JSON.stringify(body));
-	if (response?.ok) mutate(route);
+	const response = await handleWithToast(
+		putRequest(route, JSON.stringify(body))
+	);
+	if (response.ok) {
+		mutate(route);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:227]:
```suggestion
	if (response.ok) mutate(route);
```

File: frontend/util/api.ts [hunk 27]
```diff
@@ -174,14 +219,25 @@ export async function modifyDeck(
...
-	if (response?.ok) mutate(decksRoute(groupID));
+	const response = await handleWithToast(
+		deleteRequest(deckRoute(deckID)),
+		"deleteDeckToastID",
+		t`Deleting deck...`,
+		t`Deck deleted`
+	);
+	if (response.ok) {
+		mutate(decksRoute(groupID));
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:240]:
```suggestion
	if (response.ok) mutate(decksRoute(groupID));
```

File: frontend/util/api.ts [hunk 28]
```diff
@@ -190,15 +246,18 @@ export async function toggleFavorite(
...
+			favoriteDecksRoute,
+			JSON.stringify({
+				deckID: deckID,
+			})
+		)
 	);
-	if (response?.ok)
+	if (response.ok) {
 		mutateAll([decksRoute(groupID), favoriteDecksRoute, activeDecksRoute]);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:260]:
```suggestion
	if (response.ok) mutateAll([decksRoute(groupID), favoriteDecksRoute, activeDecksRoute]);
```

File: frontend/util/api.ts [hunk 29]
```diff
@@ -211,46 +270,57 @@ export async function createCard(
...
+						header: back.value,
+					},
+				],
+			})
+		)
 	);
-	if (response?.ok) mutateAll([cardsRoute(deckID), ...cardRoutes(deckID)]);
+	if (response.ok) {
+		mutateAll([cardsRoute(deckID), ...cardRoutes(deckID)]);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:290]:
```suggestion
	if (response.ok) mutateAll([cardsRoute(deckID), ...cardRoutes(deckID)]);
```

File: frontend/util/api.ts [hunk 30]
```diff
@@ -211,46 +270,57 @@ export async function createCard(
...
+			cardRoute(card.cardID),
+			JSON.stringify({
+				sides: card.sides,
+			})
+		)
 	);
-	if (response?.ok && card.deckID) mutate(cardsRoute(card.deckID));
+	if (response.ok && card.deckID) {
+		mutate(cardsRoute(card.deckID));
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:305]:
```suggestion
	if (response.ok && card.deckID) mutate(cardsRoute(card.deckID));
```

File: frontend/util/api.ts [hunk 31]
```diff
@@ -211,46 +270,57 @@ export async function createCard(
...
-	const response = await postRequest(
-		pushCardsRoute(deckID),
-		JSON.stringify({ cardID, rating })
+	const response = await handleWithToast(
+		postRequest(pushCardsRoute(deckID), JSON.stringify({ cardID, rating }))
 	);
-	if (response?.ok) mutateAll(cardRoutes(deckID));
+	if (response.ok) {
+		mutateAll(cardRoutes(deckID));
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:315]:
```suggestion
	if (response.ok) mutateAll(cardRoutes(deckID));
```

File: frontend/util/api.ts [hunk 32]
```diff
@@ -211,46 +270,57 @@ export async function createCard(
...
 	return response;
 }
 
 export async function deleteCard(deckID: string, cardID: string) {
-	const response = await deleteRequest(cardRoute(cardID));
-	if (response?.ok) mutateAll([cardsRoute(deckID), ...cardRoutes(deckID)]);
+	const response = await handleWithToast(deleteRequest(cardRoute(cardID)));
+	if (response.ok) {
+		mutateAll([cardsRoute(deckID), ...cardRoutes(deckID)]);
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:323]:
```suggestion
	if (response.ok) mutateAll([cardsRoute(deckID), ...cardRoutes(deckID)]);
```

File: frontend/util/toasts.ts [hunk 33]
```diff
@@ -0,0 +1,20 @@
...
+export async function handleWithToast(
+	promise: Promise<Response>,
+	toastID: string = "",
+	loadingMessage?: string,
+	successMessage?: string
+) {
+	if (loadingMessage) {
+		toastID = toast.loading(loadingMessage, { id: toastID });
+	}
+	const res = await promise;
```
Comment (Julian702) [inline frontend/util/toasts.ts:12]: Could you rename this variable to match the standard?
```suggestion
	const response = await promise;
```

File: frontend/util/api.ts [hunk 34]
```diff
@@ -58,19 +60,31 @@ export async function submitForm(
...
 	return response;
 }
 
 export async function deleteUser() {
-	const response = await deleteRequest(userRoute);
-	return response;
+	return await handleWithToast(
+		deleteRequest(userRoute),
+		"deleteUserToastID",
```
Comment (Julian702) [inline frontend/util/api.ts:75]: Is there a reason to sometimes using a toastID and sometimes not? I would suggest to set a custom id everywhere.

**Review by Julian702 — CHANGES_REQUESTED** [review]

File: frontend/util/api.ts [hunk 35]
```diff
@@ -141,27 +162,45 @@ export async function declineGroupInvitation(groupID: string) {
...
 export async function deleteMember(groupID: string, userID: string) {
-	const response = await deleteRequest(groupMemberRoute(groupID, userID));
-	if (response?.ok) mutate(groupMembersRoute(groupID));
+	const response = await handleWithToast(
+		deleteRequest(groupMemberRoute(groupID, userID)),
+		"deleteMemberToastID"
+	);
+	if (response.ok) {
+		mutate(groupMembersRoute(groupID));
+	}
```
Comment (Julian702) [inline frontend/util/api.ts:171]:
```suggestion
	if (response.ok) mutate(groupMembersRoute(groupID));
```

**Review by Julian702 — CHANGES_REQUESTED** [review]

File: backend/services/frontend/handler/frontend.go [hunk 36]
```diff
@@ -48,15 +48,6 @@ func (e *Frontend) RegisterHandler(c *fiber.Ctx) error {
...
 	}
-	if data.Email == "" {
-		return helper.NewFiberMissingEmailErr()
-	}
-	if data.Name == "" {
-		return helper.NewFiberMissingNameErr()
-	}
-	if data.Password == "" {
-		return helper.NewFiberMissingPasswordErr()
-	}
```
Comment (4KevR) [inline backend/services/frontend/handler/frontend.go:59]: As you removed the errors here, I think we can also remove them in the `errors.go` file

File: backend/services/frontend/handler/frontend.go [hunk 37]
```diff
@@ -104,11 +89,11 @@ func (e *Frontend) ReauthHandler(c *fiber.Ctx) error {
 	tokenString := c.Cookies("refresh_token")
 	refreshToken, err := helper.ParseJWTToken(tokenString)
 	if err != nil {
-		return helper.NewFiberUnauthorizedErr(err.Error())
+		return helper.NewFiberReauthenticateError()
```
Comment (4KevR) [inline backend/services/frontend/handler/frontend.go:92]: I just saw that the error thrown here leads to a 500 error in the response. This is due to the fact that in `main.go` of the frontend proxy, the fiber config only parses for micro errors and therefore (so my assumption) does not parse this 401 correctly. Do we want to address the issue in this PR or open a new issue for the error parsing?

**Review by 4KevR — CHANGES_REQUESTED** [review]
Just tested locally, the new toast looks awesome! Found two issues.

File: backend/services/frontend/handler/frontend.go (hunk 36 above)
Comment (TomRomeo) [reply backend/services/frontend/handler/frontend.go:59]: Good catch, I have removed the unused errors from the `errors.go` file.

File: backend/services/frontend/handler/frontend.go (hunk 37 above)
Comment (TomRomeo) [reply backend/services/frontend/handler/frontend.go:92]: You are right. I have added the feature to the error handler to also handle fiber errors. They should now also be handled correctly. Thank you for finding that! <3

**Review by 4KevR — APPROVED** [review]

**Pull request merged**